### PR TITLE
[WPE] DRM device detection does not work for imx.8 boards

### PR DESCRIPTION
**Pull request by alexgcastro**
#### 84313c5961f1b6f445504a6909c74f6ce7e25d9c
<pre>
[WPE] DRM device detection does not work for imx.8 boards
<a href="https://bugs.webkit.org/show_bug.cgi?id=270867">https://bugs.webkit.org/show_bug.cgi?id=270867</a>

Reviewed by Carlos Garcia Campos.

In case of imx.8 boards the kms device does not have the boot
VGA flag and in that situation the loop is not correctly keeping
the previous kms capable device data alive.

* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
(findDevice): Use a struct to return the fd and fileName, rework the
loop to copy and save the references even in case we traverse
all the devices. Fix name typo.
(wpeDisplayDRMConnect): Use the new struct to get the values from
the findDevice.

Canonical link: <a href="https://commits.webkit.org/276150@main">https://commits.webkit.org/276150@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/ababafdb484d0bfd7f3dc12f1b34743c065ed595

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43607 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22652 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46043 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46249 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39736 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26500 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20063 "Built successfully") | [loading-orange 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36037 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 11 flakes 217 failures 1 missing results; Uploaded test results; 34 flakes 181 failures 1 missing results; Compiled WebKit (warnings); 9 flakes 172 failures 1 missing results") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44181 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19694 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37553 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17007 "Passed tests") | 
| | [loading 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17244 "Build is in progress. Recent messages:OS: Sonoma (14.3.1), Xcode: 15.2; Running apply-patch; Checked out pull request; Running run-layout-tests-in-stress-mode; Running layout-tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1664 "Built successfully") | 
| | [loading 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39789 "Build is in progress. Recent messages:OS: Sonoma (14.3.1), Xcode: 15.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-api-tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38925 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47797 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18651 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15248 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42825 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20065 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41496 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20245 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6003 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19696 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp [hunk 1]
```diff
@@ -100,37 +104,29 @@ static std::optional<std::pair<UnixFileDescriptor, CString>> findDevice(struct u
...
-        bool isbootVGA = false;
+        bool isBootVGA = false;
         if (auto* pciDevice = udev_device_get_parent_with_subsystem_devtype(udevDevice.get(), "pci", nullptr)) {
             const char* id = udev_device_get_sysattr_value(pciDevice, "boot_vga");
-            isbootVGA = id && !strcmp(id, "1");
+            isBootVGA = id && !strcmp(id, "1");
         }
 
-        if (!isbootVGA && drmDevice)
+        if (!isBootVGA && drmDevice)
```
Comment (psaavedra) [inline Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp]: I dont see drmDevice being assigned in the new version of the code.

Comment (psaavedra) [reply Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp]: ... so this if() ia never satisfied

File: Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp [hunk 2]
```diff
@@ -100,37 +104,29 @@ static std::optional<std::pair<UnixFileDescriptor, CString>> findDevice(struct u
...
+        if (!resources || !resources->count_crtcs || !resources->count_connectors || !resources->count_encoders)
             continue;
-        }
-
-        drmDevice = WTFMove(udevDevice);
-        if (isbootVGA)
-            return std::pair<UnixFileDescriptor, CString> { WTFMove(fd), CString(filename) };
 
-        fd = { };
+        displayDevice.deviceName = CString(filename);
```
Comment (psaavedra) [inline Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp]: this change makes more clear the code. Looks good.

File: Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp (hunk 1 above)
Comment (alexgcastro) [reply Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp]: Right. I forgot to replace it, thanks!

File: Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp [hunk 3]
```diff
@@ -78,16 +78,25 @@ static void wpeDisplayDRMDispose(GObject* object)
...
+static struct DisplayDevice findDevice(struct udev* udev, const char* seatID)
 {
     RefPtr<struct udev_enumerate> udevEnumerate = adoptRef(udev_enumerate_new(udev));
     udev_enumerate_add_match_subsystem(udevEnumerate.get(), "drm");
     udev_enumerate_add_match_sysname(udevEnumerate.get(), "card[0-9]*");
     udev_enumerate_scan_devices(udevEnumerate.get());
 
-    UnixFileDescriptor fd;
-    RefPtr<struct udev_device> drmDevice;
+    struct DisplayDevice displayDevice = { };
```
Comment (carlosgcampos) [inline Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp]: No need to initialize it.

File: Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp [hunk 4]
```diff
@@ -100,37 +109,31 @@ static std::optional<std::pair<UnixFileDescriptor, CString>> findDevice(struct u
...
-        bool isbootVGA = false;
+        bool isBootVGA = false;
         if (auto* pciDevice = udev_device_get_parent_with_subsystem_devtype(udevDevice.get(), "pci", nullptr)) {
             const char* id = udev_device_get_sysattr_value(pciDevice, "boot_vga");
-            isbootVGA = id && !strcmp(id, "1");
+            isBootVGA = id && !strcmp(id, "1");
         }
 
-        if (!isbootVGA && drmDevice)
+        if (!isBootVGA && !displayDevice.drmDevice.isNull())
```
Comment (carlosgcampos) [inline Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp]:
```suggestion
        if (!isBootVGA && !displayDevice.isNull())
```

File: Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp [hunk 5]
```diff
@@ -100,37 +109,31 @@ static std::optional<std::pair<UnixFileDescriptor, CString>> findDevice(struct u
...
             continue;
-        }
 
-        drmDevice = WTFMove(udevDevice);
-        if (isbootVGA)
-            return std::pair<UnixFileDescriptor, CString> { WTFMove(fd), CString(filename) };
-
-        fd = { };
+        displayDevice.drmDevice = CString(filename);
+        displayDevice.fd = WTFMove(fd);
```
Comment (carlosgcampos) [inline Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp]:
```suggestion
        displayDevice = { CString(filename), WTFMove(fd) };
```

File: Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp [hunk 6]
```diff
@@ -238,12 +241,12 @@ static gboolean wpeDisplayDRMConnect(WPEDisplay* display, GError** error)
 
     auto session = WPE::DRM::Session::create();
     auto fdAndFilename = findDevice(udev.get(), session->seatID());
```
Comment (carlosgcampos) [inline Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp]: Let's use displayDevice now that the struct we receive is DisplayDevice, or just device

File: Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp [hunk 7]
```diff
@@ -78,16 +78,25 @@ static void wpeDisplayDRMDispose(GObject* object)
     G_OBJECT_CLASS(wpe_display_drm_parent_class)->dispose(object);
 }
 
+struct DisplayDevice {
+    CString drmDevice;
```
Comment (carlosgcampos) [inline Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp]: Device is already in the struct name, this could be just path or filename, for example

**Review by carlosgcampos — APPROVED** [review]
Much better! I have only a few minor comments

**Comment by alexgcastro** [comment]
Thanks for the review! Great additions.

**Pull request merged**